### PR TITLE
fix: Console.flix doesn't consume std in 

### LINKED ISSUE #2078: Bug in Console.readLine()
Dear Flix developers,

If you call `Console.readLine()` more than once, then it seems that all calls except the first always return `None`.

Here's a small example:

```flix
def main(_args: Array[String]): Int32 & Impure =
    println(Console.readLine());
    println(Console.readLine());
    0
```

If I run this program with the following input file, `xs.txt`,

```
1
2
3
```

then I get

```
Some(1)
None
```

but the expected result is

```
Some(1)
Some(2)
```

I'm running the program like this:

```java -jar flix-2021-07-27.jar ReadLines.flix < xs.txt```

Looking at the source https://github.com/flix/flix/blob/master/main/src/library/Console.flix#L52-L62 I notice that a new `Scanner` is created on every invocation of `Console.readLine()`. I think this is what's causing the problem: The underlying `InputStream` is not left in a state that can be used by another `Scanner`. A possible solution is to have a singleton `Scanner` that's used for every call (like Scala does, although they use a `BufferedReader` instead of a `Scanner`).

(By the way, I like the language a lot, and I'm having fun exploring it!) ☺️

### PR DESCRIPTION
**Pull request by JonathanStarup**
I am not sure what to do with the remaining code (fixes #2078).

### CODE REVIEW

File: main/src/library/Console.flix [hunk 1]
```diff
@@ -15,42 +15,70 @@
...
+            case Some(c) =>
+                print(writer(c), s)
+        }
 
     ///
-    /// Operations on the Standard Input Stream.
+    /// Wrapper for `java.lang.System.console().writer().println(s)`.
+    /// Returns `Err()` if no console is available.
     ///
-    namespace StdIn {
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:37]: Not sure why this namespace was lost.

I think we should preserve the same structure with `StdIn/StdOut` etc. If you want to change it, we can do it separately.

File: main/src/library/Console.flix [hunk 2]
```diff
@@ -15,42 +15,70 @@
...
-    /// Alias for `Console.StdIn.readLine`.
+    /// Wrapper for `java.lang.System.Console().readLine()`.
+    /// Returns `None` if a there is no line to read or no console is available.
     ///
-    pub def readLine(): Option[String] & Impure = Console/StdIn.readLine(())
+    pub def readLine(): Option[String] & Impure =
+        import java.io.Console.readLine();
+        let* c = getConsole();
+        let line = readLine(c);
+        if (isNull(line)) None else Some(line)
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:29]: Use `Object.toOption` (also in other places).

File: main/src/library/Console.flix [hunk 3]
```diff
@@ -15,42 +15,70 @@
...
+                flush(c)
+        }
+
+    ///
+    ///  Wrapper for `java.lang.System.console()` with `None` instead of null.
+    ///
+    def getConsole(): Option[##java.io.Console] & Impure =
+        import static java.lang.System.console();
+        let nullableConsole = console();
+        if (isNull(nullableConsole)) None
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:80]: Object.toOption

File: main/src/library/Console.flix [hunk 4]
```diff
@@ -15,42 +15,70 @@
...
-                None
-    }
+    ///
+    /// Wrapper for `java.lang.System.console().flush()`
+    ///
+    pub def flush(): Unit & Impure =
+        import java.io.Console.flush();
+        match getConsole() {
+            case None => ()
+            case Some(c) =>
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:70]: Move on same line, Some(c) => flush()

File: main/src/library/Console.flix [hunk 5]
```diff
@@ -15,42 +15,70 @@
...
+    /// Wrapper for `java.lang.System.console().writer().println(s)`.
+    /// Returns `Err()` if no console is available.
     ///
-    namespace StdIn {
+    pub def printLine(s: String): Unit & Impure =
+        import java.io.Console.writer();
+        import java.io.PrintWriter.println(String);
+        match getConsole() {
+            case None => ()
+            case Some(c) =>
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:59]: Move on same line.

File: main/src/library/Console.flix [hunk 6]
```diff
@@ -733,7 +761,7 @@ namespace Console {
         else if (isAnsiTerminal())
             ColorSupport.Ansi
         else
-            ColorSupport.None
+            ColorSupport.NoColor
```
Comment (magnus-madsen) [inline main/src/library/Console.flix:765]: So I guess nothing is done for StdErr because it is not supported by java's console?

File: main/src/library/Console.flix (hunk 1 above)
Comment (JonathanStarup) [reply main/src/library/Console.flix:37]: I removed it since console is now specifically not stdIn

File: main/src/library/Console.flix (hunk 6 above)
Comment (JonathanStarup) [reply main/src/library/Console.flix:767]: I was a bit in doubt of what to do with the namespaces. StdOut and StdIn works like they should. They can write to their streams. StdIn doesn't without global state seemingly so I don't know how to fix that. So this PR is more like adding access to the java console class, not really fixing StdIn.

**Comment by JonathanStarup** [comment]
New solution: homemade line reader. This is better than the previous version (could only read one line) but is not perfect. It cannot accept bot CR and CRLF as line endings because you cannot backtrack on a inputStreamReader (and using inputstream reader directly requires byte decoding).

This is meant as a band-aid. It solves the issue but can be improved

**Comment by magnus-madsen** [comment]
I am pretty scared to use such a home-made solution. There might be tons of corner cases we miss. (Emojis? Escape characters? left-to-right override?) I don't think we can do it. At least not without a battery of tests.

**Comment by magnus-madsen** [comment]
Maybe we should just return an enum that wraps `java.io.Console` and provide operations on that?

**Comment by magnus-madsen** [comment]
Or maybe we need to rethink entirely how to deal with stdin/stdout/stderr and consoles.

**Comment by JonathanStarup** [comment]
I think there should be a flix way of accessing standard in and that probably  needs static generated class.  Console could maybe be nice but I don't think it is a substitude for stdin.

**Comment by magnus-madsen** [comment]
I am closing this pending the redesign of `Console`.
